azure/src/main/java/org/apache/iceberg/azure/adlsv2/VendedAdlsCredentialProvider.java

```diff
@@ -0,0 +1,157 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.azure.adlsv2;
+
+import com.azure.core.credential.AccessToken;
+import com.azure.core.credential.SimpleTokenCache;
+import java.io.Serializable;
+import java.time.Instant;
+import java.time.ZoneOffset;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+import org.apache.commons.io.IOUtils;
+import org.apache.iceberg.azure.AzureProperties;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.rest.ErrorHandlers;
+import org.apache.iceberg.rest.HTTPClient;
+import org.apache.iceberg.rest.HTTPHeaders;
+import org.apache.iceberg.rest.RESTClient;
+import org.apache.iceberg.rest.auth.DefaultAuthSession;
+import org.apache.iceberg.rest.auth.OAuth2Properties;
+import org.apache.iceberg.rest.auth.OAuth2Util;
+import org.apache.iceberg.rest.credentials.Credential;
+import org.apache.iceberg.rest.responses.LoadCredentialsResponse;
+import org.apache.iceberg.util.SerializableMap;
+import reactor.core.publisher.Mono;
+
+public class VendedAdlsCredentialProvider implements Serializable, AutoCloseable {
+
+  public static final String URI = "credentials.uri";
+
+  private final SerializableMap<String, String> properties;
+  private transient volatile Map<String, SimpleTokenCache> azureSasCredentialMap;
+  private transient volatile RESTClient client;
+
+  public VendedAdlsCredentialProvider(Map<String, String> properties) {
+    Preconditions.checkArgument(null != properties, "Invalid properties: null");
+    Preconditions.checkArgument(null != properties.get(URI), "Invalid URI: null");
+    this.properties = SerializableMap.copyOf(properties);
+    azureSasCredentialMap = Maps.newHashMap();
+  }
+
+  public String credentialForAccount(String storageAccount) {
+    Map<String, SimpleTokenCache> tokenCacheForAccountMap = azureSasCredentialMap();
+    if (tokenCacheForAccountMap.containsKey(storageAccount)) {
+      return tokenFromCache(tokenCacheForAccountMap.get(storageAccount));
+    } else {
+      SimpleTokenCache tokenCache =
+          new SimpleTokenCache(() -> Mono.fromSupplier(() -> sasTokenSupplier(storageAccount)));
+      tokenCacheForAccountMap.put(storageAccount, tokenCache);
+      return tokenFromCache(tokenCache);
+    }
+  }
+
+  private String tokenFromCache(SimpleTokenCache simpleTokenCache) {
+    return simpleTokenCache.getToken().map(AccessToken::getToken).block();
+  }
+
+  private AccessToken sasTokenSupplier(String storageAccount) {
```

Comment:
the name says `supplier` but we don't have any supplier here. I'd suggest to rename this to `sasTokenForAccount()` or something similar